pkg/kubefed2/federate/federate.go

```diff
@@ -183,32 +185,68 @@ func GetFedResources(hostConfig *rest.Config, qualifiedTypeName, qualifiedName c
 	}, nil
 }
 
-func lookupTypeDetails(hostConfig *rest.Config, typeName ctlutil.QualifiedName) (*fedv1a1.FederatedTypeConfig, error) {
-	apiResource, err := enable.LookupAPIResource(hostConfig, typeName.Name, "")
+func getTypeConfig(hostConfig *rest.Config, typeName, namespace string, outputYAML bool) (typeconfig.Interface, error) {
+	// Lookup kubernetes API availability
+	apiResource, err := enable.LookupAPIResource(hostConfig, typeName, "")
 	if err != nil {
-		return nil, err
+		return nil, errors.Wrapf(err, "Failed to find target API resource %s", typeName)
 	}
-	typeName.Name = typeconfig.GroupQualifiedName(*apiResource)
+	glog.V(2).Infof("API Resource for %s found", typeName)
 
-	client, err := genericclient.New(hostConfig)
+	typeConfig, err := attemptToLookupFederatedType(hostConfig, *apiResource, namespace)
+	if err == nil {
+		return typeConfig, nil
+	}
+	if apierrors.IsNotFound(err) && !outputYAML {
+		return nil, errors.Errorf("%v. Try 'kubefed2 enable type %s' before federating the resource", err, typeName)
+	}
+	if outputYAML {
+		glog.V(1).Infof("Lookup of installed type failed: %v - Falling back on generated default values instead.", err)
+		return enable.GenerateTypeConfigForTarget(*apiResource, enable.NewEnableTypeDirective()), nil
+	}
+
+	return nil, err
+}
+
+func attemptToLookupFederatedType(hostConfig *rest.Config, apiResource metav1.APIResource, namespace string) (typeconfig.Interface, error) {
+	resolvedTypeName := typeconfig.GroupQualifiedName(apiResource)
+	// Lookup federated API availability
+	federatedTypeName := federatedTypeName(apiResource)
+	_, err := enable.LookupAPIResource(hostConfig, federatedTypeName, "")
```

Comment:
What is your intent in looking up the api resource of the federated type?  If the federated type were not present for whatever reason, wouldn't an appropriate error be returned when attempting to create an instance of the type?

If it is desirable to lookup the api resource in this function, is there a reason to search for a computed type name (via `federatedTypeName')?  There is no guarantee that the computed name is valid, and performing the check after retrieving the type config would allow for the configured type name to be used.

Reply:
2 reasons:
1. It provides clear indication to the user as to what is missing; its however debatable if the typeconfig could exist without the federated API CRD or vice versa, as the recommended way is to use `kubefed2 enable` or `helm install`. This is but a possibility.
2. A more pressing reason. I did observe that the create resource on a resource client for an API resource sometimes returns with error: `the server could not find the requested resource` even when the resource CRD is available; if its called immediately after the resource CRD is created. The lookup however always succeeds. Its probably has something to do with when is the resource reflected in the discovery API (for lookup) and against the calls of dynamic client in the API server.
This is further pronounced in #663 which implements `--enable-type`, where I have implemented a retry loop with a short timeout (in case the lookup succeeded but create returns `NotFound` error) in create to ensure this is handled correctly.

Having stated above I could further think of a possible scenario, where an user wants to use a precreated typeconfig (probably to override target `group/version` ), haven't used `kubefed2 enable` so the federated API resource is not available; creates typeconfig only to convert resources (-o yaml). This code would still default to a generated type if lookup is kept here.
I am not sure if this is an ideal scenario, but probably is a possibility.
For this possibility to work cleanly and still serve the reasons 1 and 2 above, moving The lookup  of federated type immediately before create resource. The retry loop is introduced in #663.

Reply:
EDIT: On second observations, I was wrong on `Lookup()` always succeeding, as I saw a failure on `Lookup()` too sometimes when running with --enable-type. I m removing the `Lookup()` completely here and relying on failure reported from create resource alone.